Add approx_percentile forms with accuracy

jagill opened the pull request:
Previously you could only specify an accuracy to approx_percentile if
you also specified weights.  This adds an accuracy option for all forms.

```
== RELEASE NOTES ==

General Changes
* Add forms of approx_percentile accepting an accuracy parameter.

```

jagill commented:
cc @aweisberg

aweisberg commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -244,6 +258,12 @@ Approximate Aggregate Functions
+
+    As ``approx_percentile(x, w, percentages)``, but with a maximum rank error of
+    ``accuracy``. ``accuracy`` must be a value greater than zero and less than
+    one, and it must be constant for all input rows.
```
What does it mean for it to be constant for all input rows?

jagill replied:
I actually copy-pasted this from the other descriptions, but it means this:
The aggregation function actually takes a full set of parameters for each row being aggregated.  Thus the quantity, weight, etc are taken from each row.  But actually, so are the percentile (or percentile array) and accuracy parameters!  However, if in the query, you just put a constant/literal value (like 0.99 for percentile, or 0.01 for accuracy), presto expands this and feeds those constant values in each row.  you could calculate it based on the row!  But that would be confusing and meaningless.

Would it be better to say "accuracy must be a constant value greater than zero and less than one"?

aweisberg replied:
I think so because a naive reader like myself couldn't even imagine that you could feed it the output of an expression like say random.

aweisberg commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateLongPercentileAggregations.java:
```diff
@@ -32,19 +32,37 @@
 
     @InputFunction
-    public static void input(@AggregationState DigestAndPercentileState state, @SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.DOUBLE) double percentile)
+    public static void input(
```
Could all these input functions just call weightedInput with a weight of 1.0? It looks like in QuantileDigest add without a weight is just add with a weight of 1. Then a lot of duplicated code could be removed.

jagill replied:
We could!  We'd end up checking the weight every row (checking that `weight > 0`).  Do we have a feel for the performance implications?  It's worth noting that in the case of accuracy, the check is only done once (when the digest is created).

jagill replied:
I suppose we could have a private helper method that didn't check, and everything delegate to that.

aweisberg replied:
Sounds like you have a plan :-)

aweisberg commented on presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestApproximatePercentileAggregation.java:
```diff
@@ -40,22 +40,27 @@
+    private static final InternalAggregationFunction FLOAT_APPROXIMATE_PERCENTILE_AGGREGATION_WITH_ACCURACY = getAggregation(REAL, DOUBLE, DOUBLE);
     private static final InternalAggregationFunction FLOAT_APPROXIMATE_PERCENTILE_WEIGHTED_AGGREGATION = getAggregation(REAL, BIGINT, DOUBLE);
     private static final InternalAggregationFunction FLOAT_APPROXIMATE_PERCENTILE_WEIGHTED_AGGREGATION_WITH_ACCURACY = getAggregation(REAL, BIGINT, DOUBLE, DOUBLE);
+
```
Intentional additional line break?

jagill replied:
Yes.  The other types (double, long) have the array and non-array sections separated by a line break, so I thought I'd add one here for consistency.  Happy to remove if desired.

aweisberg replied:
Sounds good, keep it.

jagill commented:
addressed comments cc @aweisberg

jagill commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -208,14 +208,26 @@ Approximate Aggregate Functions
-    given ``percentage``. The value of ``percentage`` must be between zero and
-    one and must be constant for all input rows.
+    given ``percentage``. ``percentage`` must be a constant value between zero
+    and one.
```
I won't hate you!  But if you have wording that addresses your concerns, it may be easier for me to put that in.

aweisberg commented:
We will need to add a committer to get this in. +Rebecca.

jagill commented:
@aweisberg:  Tracking down the error paths for the EsriGeometrySerde, they are logically possible but would be caught earlier.  I've removed that.

For the GeometryType code, originally the issue was that after some errors, there'd be a debug printout that would error out.  However, I can't reproduce that now, so I've removed that code and if it comes up again I'll replace the code and put in a test case that tests it.
